File: 27. Print all root to leaf paths.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
void paths(Node* root, vector<int>& cur)
{
    if (!root)
        return;
    cur.push_back(root->data);
    if (!root->left && !root->right)
    {
        for (int v : cur) cout << v << " ";
        cout << "\n";
    }
    else
    {
        paths(root->left, cur);
        paths(root->right, cur);
    }
    cur.pop_back();
}
int main()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    vector<int> cur;
    paths(root, cur);
    return 0;
}

File: 27. Print all root to leaf paths.py
class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def print_paths(root):
    def helper(node,path):
        if not node: return
        path.append(node.data)
        if not node.left and not node.right:
            print(' '.join(map(str,path)))
        else:
            helper(node.left,path); helper(node.right,path)
        path.pop()
    helper(root,[])

if __name__=='__main__':
    root=Node(1); root.left=Node(2); root.right=Node(3); root.left.left=Node(4)
    print_paths(root)
